mllaunchpad/cli.py

```diff
@@ -15,104 +17,117 @@
+)
+@pass_settings
+def main(settings, log_config, config, verbose):
+    """Train, test or run a config file's model."""
+    # Initialize logging before any library code so that mllp can log stuff
+    if log_config:
+        settings.logger = logutil.init_logging(log_config, verbose=verbose)
+    else:
+        settings.logger = logutil.init_logging(verbose=verbose)
+    settings.verbose = verbose
+    settings.conf_file = config
+
+
+@main.command()
+@pass_settings
+def train(settings):
+    """Run training, store created model and metrics."""
+    _, metrics = lp.train_model(settings.config)
+    print(metrics)
+
+
+@main.command()
+@pass_settings
+def retest(settings):
+    """Retest existing model, update metrics."""
+    metrics = lp.retest(settings.config)
+    print(metrics)
+
+
+@main.command()
+@pass_settings
+def api(settings):
+    """Run API server in unsafe debug mode."""
+    settings.logger.warning(
+        "Starting Flask debug server. In production, please "
+        "use a WSGI server, e.g.\n"
+        "'export LAUNCHPAD_CFG=addition_cfg.yml'\n"
+        "'gunicorn -w 4 -b 127.0.0.1:5000 mllaunchpad.wsgi:application'"
     )
-    if cmd == "train":
-        _, _ = lp.train_model(conf)
-    elif cmd == "retest":
-        _ = lp.retest(conf)
-    elif cmd == "predict":
-        # TODO decide: get batch args from config? Don't support arguments?
-        _ = lp.predict(conf, arg_dict={})
-    elif cmd == "api":
-        logger.warning(
-            "Starting Flask debug server. In production, please "
-            "use a WSGI server, e.g. "
-            "'export LAUNCHPAD_CFG=addition_cfg.yml'"
-            "'gunicorn -w 4 -b 127.0.0.1:5000 mllaunchpad.wsgi:application'"
-        )
-        app = Flask(__name__, root_path=conf["api"].get("root_path"))
-        ModelApi(conf, app)
-        # Flask apps must not be run in debug mode in production, because this allows for arbitrary code execution.
-        # We know that and advise the user that this is only for debugging, so this is not a security issue (marked nosec):
-        app.run(debug=True)  # nosec
-    elif cmd == "genraml":
-        print(generate_raml(conf, data_source_name=raml_ds))
+    app = Flask(__name__, root_path=settings.config["api"].get("root_path"))
+    ModelApi(settings.config, app)
+    # Flask apps must not be run in debug mode in production, because this allows for arbitrary code execution.
+    # We know that and advise the user that this is only for debugging, so this is not a security issue (marked nosec):
+    app.run(debug=True)  # nosec
+
+
+@main.command()
+@click.argument("json-file", type=click.File("r"))
+@pass_settings
+def predict(settings, json_file):
+    """Run prediction on features from JSON file ( - for stdin).
+
+    \b
+    Example JSON: { "petal.width": 1.4, "petal.length": 2.0,
+                    "sepal.width": 1.8, "sepal.length": 4.0 }
+    """
```

Comment:
Do we always have to provide json with the arguments? Is there a possibility to add some arguments in for a single prediction api call after '?'. (as previous comment it is just a question for the general understanding not a remark)

Reply:
Keep in mind that this PR only deals with the command line interface, so it won’t change how API calls work. The ‘mllaunchpad predict’ command calls prediction functionality directly without going through the Web API. This is also why there is no ‘?’ In the command line interface.

Before this change, it was not possible at all to provide any input to the prediction from the command line. Allowing for providing JSON from a file (or stdin) was the easiest way to add this functionality (as prediction functionality expects the features as a dictionary). I’d wait and see if and how people use the new possibilities that ‘mllaunchpad predict’ provides, and use what we learned to decide whether we need to support more direct command line string format (which could or could not look like query parameters) of prediction input.

Again, just to be unambiguous: we’re not taking away any ways that one was able to try out their models up until now (running the web api, then calling its url via browser/curl/postman), bit we’re adding an additional way to do prediction without going through the Web API